subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -88,47 +104,55 @@ public void registerWatchableHierarchy(File watchableHierarchy, SnapshotHierarch
         WatchableHierarchies.Invalidator invalidator = (location, currentRoot) -> currentRoot.invalidate(location, SnapshotHierarchy.NodeDiffListener.NOOP);
         SnapshotHierarchy newRoot = watchableHierarchies.removeWatchedHierarchiesOverLimit(
-            root, watchedHierarchies::contains, invalidator
+            root,
+            watchedHierarchy -> watchedHierarchies.contains(watchedHierarchy.toString()),
+            invalidator
         );
         newRoot = watchableHierarchies.removeUnwatchedSnapshots(
             newRoot,
             invalidator
         );
 
         updateWatchedHierarchies(newRoot);
-        LOGGER.info("Watched directory hierarchies: {}", watchedHierarchies);
+        LOGGER.info("Watched directory hierarchies: {}", watchedRoots);
         return newRoot;
     }
 
     @Override
     public int getNumberOfWatchedHierarchies() {
-        return watchedHierarchies.size();
+        return watchedRoots.size();
     }
 
     private void updateWatchedHierarchies(SnapshotHierarchy root) {
-        Set<String> snapshotsAlreadyCoveredByOtherHierarchies = new HashSet<>();
+        AtomicReference<FileHierarchySet> newWatchedHierarchiesReference = new AtomicReference<>(DefaultFileHierarchySet.of());
         Set<Path> hierarchiesWithSnapshots = watchableHierarchies.getWatchableHierarchies().stream()
             .flatMap(watchableHierarchy -> {
+                FileHierarchySet newWatchableHierarchies = newWatchedHierarchiesReference.get();
+                if (newWatchableHierarchies.contains(watchableHierarchy.toString())) {
+                    return Stream.empty();
+                }
                 CheckIfNonEmptySnapshotVisitor checkIfNonEmptySnapshotVisitor = new CheckIfNonEmptySnapshotVisitor(watchableHierarchies);
-                root.visitSnapshotRoots(watchableHierarchy.toString(), new FilterAlreadyCoveredSnapshotsVisitor(checkIfNonEmptySnapshotVisitor, snapshotsAlreadyCoveredByOtherHierarchies));
+                root.visitSnapshotRoots(watchableHierarchy.toString(), new FilterAlreadyCoveredSnapshotsVisitor(checkIfNonEmptySnapshotVisitor, newWatchableHierarchies));
                 if (checkIfNonEmptySnapshotVisitor.isEmpty()) {
                     return Stream.empty();
                 }
+                newWatchedHierarchiesReference.updateAndGet(watchableHierarchies -> watchableHierarchies.plus(watchableHierarchy.toFile()));
                 return checkIfNonEmptySnapshotVisitor.containsOnlyMissingFiles()
                     ? Stream.of(locationOrFirstExistingAncestor(watchableHierarchy))
                     : Stream.of(watchableHierarchy);
             })
             .collect(Collectors.toSet());
+        watchedHierarchies = newWatchedHierarchiesReference.get();
 
-        Set<Path> newWatchedHierarchies = resolveHierarchiesToWatch(hierarchiesWithSnapshots);
+        Set<Path> newWatchedRoots = resolveHierarchiesToWatch(hierarchiesWithSnapshots);
```

Comment:
Is it possible to have duplicates? In other words, is there a need for a `Set`?

Can we maybe pass the stream to this method instead of collecting it to a set first?

Reply:
> Is it possible to have duplicates? In other words, is there a need for a Set?

You mean in `hierarchiesWithSnapshots`? Now it isn't possible any more, so we could pass a Stream to `resolveHierarchiesToWatch()`.

I also wonder if it would be possible to extract the `newWatchedRoots` from the `watchedHierarchies` aka the FileHierarchySet directly. All the information should be in there already.

Reply:
Yeah, that would allow us to do away with the `AbstractReference`, too. Let's publish a snapshot before doing that, though.

Reply:
See https://github.com/gradle/gradle/issues/14183 for obtaining the set of `Path`s from the `FileHierarchySet`.